File: src/test/java/com/technews/crawling/SpringBlogCrawlingTest.java
package com.technews.crawling;

import lombok.Builder;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class SpringBlogCrawlingTest {

    private static final String CATEGORY_URL = "https://spring.io/blog/category/";
    private static final String BLOG_URL = "https://spring.io";
    private static final String ENGINEERING = "engineering";
    private static final String RELEASES = "releases";
    private static final String NEWS = "news";

    @Test
    void test() throws IOException {
        System.out.println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> ENGINEERING");
        final List<Post> engineeringPosts = getPostInfo(ENGINEERING);
        engineeringPosts.stream().forEach(System.out::println);

        System.out.println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> RELEASES");
        final List<Post> releasesPosts = getPostInfo(RELEASES);
        releasesPosts.stream().forEach(System.out::println);

        System.out.println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> NEWS");
        final List<Post> newsPosts = getPostInfo(NEWS);
        newsPosts.stream().forEach(System.out::println);
    }

    private static List<Post> getPostInfo(final String category) throws IOException {
        final List<Post> result = new ArrayList<>();
        final Document doc = Jsoup.connect(CATEGORY_URL + category).get();

        final Elements elements = doc.select(".blog-post");
        for (Element element : elements) {
            final String title = getTitle(element);
            final Meta meta = getMeta(element);
            final String url = getPostUrl(element);

            result.add(Post.builder()
                    .title(title)
                    .meta(meta)
                    .url(url)
                    .build());
        }

        return result;
    }

    private static String getTitle(final Element element) {
        try {
            return element.select(".has-text-weight-medium").text();
        } catch (Exception e) {
            return StringUtils.EMPTY;
        }
    }

    private static Meta getMeta(final Element element) {
        final String meta = element.select(".meta").text();
        final String[] metas = meta.split("\\|");
        if (metas.length == 0) {
            return Meta.EMPTY;
        }

        try {
            return Meta.builder()
                    .category(metas[0])
                    .writer(metas[1])
                    .date(metas[2])
                    .build();
        } catch (Exception e) {
            return Meta.EMPTY;
        }
    }

    private static String getPostUrl(final Element element) {
        try {
            final String url = element.select(".button").attr("href");
            return BLOG_URL + url;
        } catch (Exception e) {
            return StringUtils.EMPTY;
        }
    }

    private record Post(
            String title,
            Meta meta,
            String url
    ) {
        @Builder
        private Post {
        }
    }

    private record Meta(
            String category,
            String writer,
            String date
    ) {
        public static final Meta EMPTY = new Meta(StringUtils.EMPTY, StringUtils.EMPTY, StringUtils.EMPTY);

        @Builder
        private Meta {
        }
    }
}

File: src/test/java/com/technews/crawling/blog/SpringBlogCrawlingTest.kt
package com.technews.crawling.blog

import io.kotest.core.spec.style.BehaviorSpec
import mu.KotlinLogging
import org.jsoup.Jsoup

private val logger = KotlinLogging.logger {}

class SpringBlogCrawlingTest : BehaviorSpec({
    Given("Spring BLog 정보 크롤링") {
        val categories = listOf(ENGINEERING, RELEASES, NEWS)

        categories.forEach { category ->
            println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> $category")
            val posts = fetchPostsByCategory(category)
            posts.forEach { println(it) }
        }
    }
}) {
    companion object {
        private const val CATEGORY_URL = "https://spring.io/blog/category/"
        private const val BLOG_URL = "https://spring.io"
        private const val ENGINEERING = "engineering"
        private const val RELEASES = "releases"
        private const val NEWS = "news"

        private fun fetchPostsByCategory(category: String): List<Post> {
            return runCatching {
                Jsoup.connect("$CATEGORY_URL$category").get()
                    .select(".blog-post")
                    .mapNotNull { element ->
                        val title = element.select(".has-text-weight-medium").text().ifEmpty { null }
                        val meta = parseMeta(element.select(".meta").text())
                        val url =
                            element.select(".button").attr("href").takeIf { it.isNotEmpty() }?.let { BLOG_URL + it }

                        if (title != null && meta != null && url != null) {
                            Post(title, meta, url)
                        } else {
                            null
                        }
                    }
            }.getOrElse {
                logger.error("Failed to fetch posts for category: $category, error: ${it.message}", it)
                emptyList()
            }
        }

        private fun parseMeta(metaText: String): Meta? {
            val metaParts = metaText.split("|").map { it.trim() }
            return if (metaParts.size >= 3) {
                Meta(metaParts[0], metaParts[1], metaParts[2])
            } else {
                logger.warn("Invalid meta format: $metaText")
                null
            }
        }

        private data class Post(
            val title: String,
            val meta: Meta,
            val url: String,
        )

        private data class Meta(
            val category: String,
            val writer: String,
            val date: String,
        )
    }
}
